grails-app/services/org/pih/warehouse/data/ProductAssociationDataService.groovy

```diff
@@ -0,0 +1,137 @@
+                )
+                return
+            }
+            if (!product.active) {
+                command.errors.reject(
+                    "Row ${index + 1}: Product with code '${productAssociationInstance.product?.id}' is inactive"
+                )
+                return
+            }
+            if (!associatedProduct.active) {
+                command.errors.reject(
+                    "Row ${index + 1}: Product with code '${productAssociationInstance.associatedProduct?.id}' is inactive"
+                )
+                return
+            }
+            if (product == associatedProduct) {
+                command.errors.reject("Cannot associate a product with itself")
+                return
+            }
+
+            // Check for association duplicates
+            List<ProductAssociation> foundProductAssociations = ProductAssociation.findAllWhere([
+                    product             : productAssociationInstance.product,
+                    associatedProduct   : productAssociationInstance.associatedProduct,
+                    code                : productAssociationInstance.code,
+            ])
+            if (foundProductAssociations && foundProductAssociations.size() > 0) {
+                command.errors.reject("Row ${index + 1}: Association already exists")
+                return
+            }
+
+
+            // find two-way association pairs
+            def twoWayAssociationPair = command.data.find {
+                it['product.id'] == params['associatedProduct.id'] &&
+                it['associatedProduct.id'] == params['product.id'] &&
+                it['code'] == params['code']
+            }
+
+            if (twoWayAssociationPair) {
+                // mark this association as two-way association
+                params.hasMutualAssociation = true
+                if (twoWayAssociationPair['quantity'] * productAssociationInstance.quantity != 1) {
+                    command.errors.reject(
+                            "Row ${index + 1}: Quantity of Product association for Products with codes " +
+                            "'${productAssociationInstance.product?.id}' and " +
+                            "'${twoWayAssociationPair['associatedProduct.id']}' does not match"
+                    )
+                    return
+                }
+            }
+
+            if (!productAssociationInstance.validate()) {
+                productAssociationInstance.errors.each { BeanPropertyBindingResult error ->
+                    command.errors.reject(
+                            "Row ${index + 1}: Product Association with Product code " +
+                            "'${productAssociationInstance.product?.id}': ${error.getFieldError()}"
+                    )
+                }
+            }
+        }
+    }
+
+    void importData(ImportDataCommand command) {
```

Comment:
Let's add more comments to make this method more readable. I'll review again once the comments are there.